File: internal/filetools/filetools_test.go
package filetools

import (
	"archive/zip"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func TestIsText(t *testing.T) {
	if !IsText([]byte("plain text\nline two")) {
		t.Error("plain text should be text")
	}
	if IsText([]byte("bin\x00ary")) {
		t.Error("NUL byte should mark binary")
	}
	// Large text with no NUL is text (only the probe is inspected).
	if !IsText([]byte(strings.Repeat("a", 100_000))) {
		t.Error("large no-NUL blob should be text")
	}
}

func TestDetectFormat(t *testing.T) {
	cases := map[string]string{
		"a.lrc": "lrc", "b.srt": "srt", "c.vtt": "vtt",
		"d.ass": "ass", "e.ssa": "ass", "f.ttml": "ttml",
		"g.md": "markdown", "h.markdown": "markdown", "i.txt": "text", "j": "text",
	}
	for path, want := range cases {
		if got := DetectFormat(path, nil); got != want {
			t.Errorf("DetectFormat(%q) = %q, want %q", path, got, want)
		}
	}
}

func TestProtectedAndVerify(t *testing.T) {
	orig := "[00:01.00]Line one\n[00:02.50]Line two\n[01:00.00]Line three"
	good := "[00:01.00]첫 줄\n[00:02.50]둘째 줄\n[01:00.00]셋째 줄"
	bad := "[00:01.00]첫 줄\n[00:03.00]둘째 줄\n[01:00.00]셋째 줄" // middle timestamp changed
	if err := VerifyProtected("lrc", orig, good); err != nil {
		t.Errorf("good lrc should verify: %v", err)
	}
	if err := VerifyProtected("lrc", orig, bad); err == nil {
		t.Error("changed timestamp must fail verification")
	}
	// Plain text has no protected lines and always verifies.
	if err := VerifyProtected("text", "hello", "bonjour"); err != nil {
		t.Errorf("text should always verify: %v", err)
	}
}

func TestProtectedLinesAdditionalFormats(t *testing.T) {
	srt := "1\n00:00:01,000 --> 00:00:02,000\nHello\n"
	vtt := "WEBVTT\n\n00:01.000 --> 00:02.000\nalign:start\nHi\n"
	ass := "[Script Info]\nTitle: T\nFormat: Layer, Start, End, Style, Name, MarginL, MarginR, MarginV, Effect, Text\nDialogue: 0,0:00:01.00,0:00:02.00,Default,,0,0,0,,Hello\n"
	ttml := `<tt><body><p begin="0s">Hi</p></body></tt>`

	if got := ProtectedLines("srt", []byte(srt)); len(got) != 1 || !strings.Contains(got[0], "-->") {
		t.Errorf("srt protected = %v", got)
	}
	v := ProtectedLines("vtt", []byte(vtt))
	if len(v) != 2 || v[0] != "WEBVTT" {
		t.Errorf("vtt protected = %v", v)
	}
	a := ProtectedLines("ass", []byte(ass))
	if len(a) != 4 {
		t.Errorf("ass protected = %v", a)
	}
	if !strings.Contains(a[3], "Dialogue: 0,0:00:01.00,0:00:02.00,Default,,0,0,0,,") {
		t.Errorf("ass dialogue prefix not kept: %q", a[3])
	}
	if got := ProtectedLines("ttml", []byte(ttml)); len(got) != 6 {
		t.Errorf("ttml protected = %v", got)
	}
}

func TestReadEpub(t *testing.T) {
	dir := t.TempDir()
	f := filepath.Join(dir, "book.epub")
	if err := writeTestEPUB(f); err != nil {
		t.Fatal(err)
	}
	text, err := ReadEpub(f)
	if err != nil {
		t.Fatalf("ReadEpub: %v", err)
	}
	for _, want := range []string{"Chapter one content", "Chapter two content"} {
		if !strings.Contains(text, want) {
			t.Errorf("EPUB text missing %q:\n%s", want, text)
		}
	}
	if strings.Contains(text, "<p>") || strings.Contains(text, "<html") {
		t.Errorf("markup not stripped:\n%s", text)
	}
}

// writeTestEPUB builds a minimal EPUB: container.xml → OPF → two XHTML
// chapters.
func writeTestEPUB(path string) error {
	ch1 := "<html><body><h1>One</h1><p>Chapter one content &amp; stuff</p></body></html>"
	ch2 := "<html><body><p>Chapter two content</p></body></html>"
	files := map[string]string{
		"META-INF/container.xml": `<?xml version="1.0"?><container><rootfiles><rootfile full-path="OEBPS/content.opf" media-type="application/oebps-package+xml"/></rootfiles></container>`,
		"OEBPS/content.opf": `<?xml version="1.0"?><package><manifest>
			<item id="c1" href="ch1.xhtml" media-type="application/xhtml+xml"/>
			<item id="c2" href="ch2.xhtml" media-type="application/xhtml+xml"/>
			</manifest><spine>
			<itemref idref="c1"/><itemref idref="c2"/>
			</spine></package>`,
		"OEBPS/ch1.xhtml": ch1,
		"OEBPS/ch2.xhtml": ch2,
	}
	out, err := os.Create(path)
	if err != nil {
		return err
	}
	defer out.Close()
	zw := zip.NewWriter(out)
	for name, content := range files {
		w, err := zw.Create(name)
		if err != nil {
			return err
		}
		if _, err := w.Write([]byte(content)); err != nil {
			return err
		}
	}
	return zw.Close()
}

File: test/filetools.test.ts
import { strict as assert } from "node:assert";
import { test } from "node:test";
import { writeFileSync } from "node:fs";
import { join } from "node:path";
import { mkdtempSync } from "node:fs";
import { tmpdir } from "node:os";
import { zipSync } from "fflate";
import {
	assDialoguePrefix,
	detectFormat,
	isText,
	protectedLines,
	stripHTML,
	verifyProtected,
	readEpub,
} from "../src/core/filetools/index.js";

test("isText: NUL bytes in the probe mean binary", () => {
	assert.equal(isText(Buffer.from("hello")), true);
	assert.equal(isText(Buffer.concat([Buffer.from("hello"), Buffer.from([0])])), false);
	// NUL beyond the probe window does not matter.
	assert.equal(isText(Buffer.concat([Buffer.alloc(9000, 0x61), Buffer.from([0])])), true);
});

test("detectFormat by extension", () => {
	assert.equal(detectFormat("a.lrc", new Uint8Array()), "lrc");
	assert.equal(detectFormat("a.SRT", new Uint8Array()), "srt");
	assert.equal(detectFormat("a.vtt", new Uint8Array()), "vtt");
	assert.equal(detectFormat("a.ass", new Uint8Array()), "ass");
	assert.equal(detectFormat("a.ssa", new Uint8Array()), "ass");
	assert.equal(detectFormat("a.ttml", new Uint8Array()), "ttml");
	assert.equal(detectFormat("a.md", new Uint8Array()), "markdown");
	assert.equal(detectFormat("a.markdown", new Uint8Array()), "markdown");
	assert.equal(detectFormat("a.txt", new Uint8Array()), "text");
	assert.equal(detectFormat("noext", new Uint8Array()), "text");
});

test("protected lines: srt timing lines", () => {
	const got = protectedLines("srt", "1\n00:00:01,000 --> 00:00:02,000\nHello\n\n2\n00:00:03,000 --> 00:00:04,000\nBye\n");
	assert.deepEqual(got, ["00:00:01,000 --> 00:00:02,000", "00:00:03,000 --> 00:00:04,000"]);
	assert.deepEqual(protectedLines("text", "anything"), []);
});

test("protected lines: vtt includes the WEBVTT header and NOTEs", () => {
	const got = protectedLines("vtt", "WEBVTT\n\nNOTE a comment\n\n00:01.000 --> 00:02.000\nHello\n");
	assert.deepEqual(got, ["WEBVTT", "NOTE a comment", "00:01.000 --> 00:02.000"]);
});

test("protected lines: lrc timecodes in order", () => {
	const got = protectedLines("lrc", "[00:12.00]First line\n[01:02]Second\n[02:03.456]Third\n");
	assert.deepEqual(got, ["[00:12.00]", "[01:02]", "[02:03.456]"]);
});

test("protected lines: ass dialogue prefix and raw structure", () => {
	const ass = "[Script Info]\nTitle: x\n\n[Events]\nFormat: Layer, Start\nDialogue: 0,0:00:01.00,0:00:02.00,Default,,0,0,0,,Hello\n";
	const got = protectedLines("ass", ass);
	assert.deepEqual(got, ["[Script Info]", "Title: x", "[Events]", "Format: Layer, Start", "Dialogue: 0,0:00:01.00,0:00:02.00,Default,,0,0,0,,"]);
	assert.equal(
		assDialoguePrefix("Dialogue: 0,0:00:01.00,0:00:02.00,Default,,0,0,0,,Hello, world"),
		"Dialogue: 0,0:00:01.00,0:00:02.00,Default,,0,0,0,,",
	);
});

test("protected lines: ttml tag sequence", () => {
	const got = protectedLines("ttml", '<tt><body><p>Hi</p><p p2="1">Yo</p></body></tt>');
	assert.deepEqual(got, ["<tt>", "<body>", "<p>", "</p>", '<p p2="1">', "</p>", "</body>", "</tt>"]);
});

test("verifyProtected: identical passes, changes fail", () => {
	const srt = "1\n00:00:01,000 --> 00:00:02,000\nHello\n";
	assert.equal(verifyProtected("srt", srt, "1\n00:00:01,000 --> 00:00:02,000\nHola\n"), undefined);
	const err = verifyProtected("srt", srt, "1\n00:00:01,500 --> 00:00:02,000\nHola\n");
	assert.ok(err instanceof Error);
	assert.match(err.message, /protected line 1 changed/);
	const count = verifyProtected("srt", srt, "1\n00:00:01,000 --> 00:00:02,000\nHola\n\n2\n00:00:03,000 --> 00:00:04,000\nBye\n");
	assert.ok(count instanceof Error);
	assert.match(count!.message, /protected line count changed \(1 → 2\)/);
});

test("stripHTML: tags, scripts and entities", () => {
	assert.equal(stripHTML("<p>Hello &amp; <b>world</b></p>"), "Hello & world");
	assert.equal(stripHTML("<script>bad()</script>ok"), "ok");
	assert.equal(stripHTML("a &lt; b &gt; c &quot;d&quot; &#39;e&apos; f&nbsp;g"), 'a < b > c "d" \'e\' f g');
	assert.equal(stripHTML("4 &lt; 5 &unknown; x"), "4 < 5 &unknown; x");
	assert.equal(stripHTML("loose & amp"), "loose & amp");
});

test("readEpub: spine order, markup stripping, cap", () => {
	const container = `<?xml version="1.0"?><container><rootfiles><rootfile full-path="OEBPS/content.opf" media-type="application/oebps-package+xml"/></rootfiles></container>`;
	const opf = `<?xml version="1.0"?><package xmlns="http://www.idpf.org/2007/opf"><manifest>
<item id="ch1" href="ch1.xhtml" media-type="application/xhtml+xml"/>
<item id="ch2" href="ch2.xhtml" media-type="application/xhtml+xml"/>
<item id="cover" href="cover.xhtml" media-type="application/xhtml+xml"/>
</manifest><spine><itemref idref="ch1"/><itemref idref="ch2"/></spine></package>`;
	const ch1 = "<html><body><p>First &amp; chapter</p></body></html>";
	const ch2 = "<html><body><p>Second chapter</p><!-- note --><script>no()</script></body></html>";
	const zip = zipSync({
		"META-INF/container.xml": Buffer.from(container),
		"OEBPS/content.opf": Buffer.from(opf),
		"OEBPS/ch1.xhtml": Buffer.from(ch1),
		"OEBPS/ch2.xhtml": Buffer.from(ch2),
		"OEBPS/cover.xhtml": Buffer.from("<p>cover</p>"),
	});
	const dir = mkdtempSync(join(tmpdir(), "dscli-"));
	const path = join(dir, "book.epub");
	writeFileSync(path, Buffer.from(zip));
	assert.equal(readEpub(path), "First & chapterSecond chapter"); // spine order, cover excluded
});

test("readEpub: missing container fails loudly", () => {
	const zip = zipSync({ "a.txt": Buffer.from("x") });
	const dir = mkdtempSync(join(tmpdir(), "dscli-"));
	const path = join(dir, "bad.epub");
	writeFileSync(path, Buffer.from(zip));
	assert.throws(() => readEpub(path), /missing META-INF\/container\.xml/);
});
